src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/eventlog/EventLogService.kt

```
@@ -0,0 +1,24 @@
+package uk.gov.justice.digital.hmpps.personrecord.service.eventlog
+
+import com.microsoft.applicationinsights.TelemetryClient
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.EventLogEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.entity.PersonKeyEntity
+import uk.gov.justice.digital.hmpps.personrecord.jpa.repository.EventLogRepository
+import java.util.UUID
+
+@Component
+class EventLogService(
+  private val eventLogRepository: EventLogRepository,
+  private val telemetryClient: TelemetryClient,
+) {
+
+  fun logEvent(
+    personEntity: PersonEntity,
+    eventType: CPRLogEvents,
+    personKeyEntity: PersonKeyEntity? = null,
+  ): EventLogEntity = eventLogRepository.save(EventLogEntity.from(personEntity, eventType, getOperationId(), personKeyEntity = personKeyEntity))
+
+  private fun getOperationId() = telemetryClient.context.operation.id ?: UUID.randomUUID().toString()
```

Do we want to default to a random UUID? If for whatever reason we don't have an operationID, I think it might be as well to indicate this. It would save us the bother of trying to look it up in AppInsights and wonder why we don't find anything

`OPERATION_ID_UNAVAILABLE` or something similar might be a better bet

---

Updated 👍